### ISSUE TITLE
NumberFormatException while parsing Date in WmiUtil

### ISSUE THREAD

**Issue by svilenvul** [dblock/oshi#177]
I have noticed the following problem. When I call:
```
        SystemInfo systemInfo = new SystemInfo();
        HardwareAbstractionLayer hal = systemInfo.getHardware();
        operatingSystem = systemInfo.getOperatingSystem();
        displays = hal.getDisplays();
        fileStores = hal.getFileStores();
        memory = hal.getMemory();
        powerSources = hal.getPowerSources();
        cpu = hal.getProcessor();
        sensors = hal.getSensors();
        networks = hal.getNetworkIFs();
        drives = hal.getDiskStores();
        if (cpu != null) {
            processes = cpu.getProcesses();
        } else {
            throw new NullPointerException("Can not get processs information, because cpu info is missing !");
        }

```

I get the following error message - 

```
12:30:15.386 [ESH-safeCall-4] DEBUG o.s.o.windows.WindowsOSVersionInfoEx - Initialized OSVersionInfoEx
12:30:15.387 [main] DEBUG o.e.s.c.c.r.AbstractManagedProvider - Added new element systeminfo_computer_work_cpu_load -> systeminfo:computer:work:cpu#load to ManagedItemChannelLinkProvider.
12:30:15.387 [main] DEBUG o.e.s.c.c.r.AbstractManagedProvider - Added new element systeminfo:computer:work to ManagedThingProvider.
12:30:15.388 [ESH-safeCall-4] DEBUG o.h.platform.windows.WindowsDisplay - Initialized WindowsDisplay
12:30:15.389 [ESH-safeCall-4] DEBUG o.h.platform.windows.WindowsDisplay - Initialized WindowsDisplay
12:30:15.594 [ESH-safeCall-4] DEBUG o.h.p.windows.WindowsPowerSource - Initialized WindowsPowerSource
12:30:15.594 [ESH-safeCall-4] DEBUG o.h.common.AbstractCentralProcessor - Oracle MXBean detected.
12:30:15.601 [ESH-safeCall-5] DEBUG o.h.p.w.WindowsCentralProcessor - Tick counter queries added.  Initializing with first query.
12:30:15.605 [ESH-safeCall-5] DEBUG o.h.p.w.WindowsCentralProcessor - Initialized Processor
12:30:15.606 [ESH-safeCall-4] DEBUG o.h.p.w.WindowsCentralProcessor - Tick counter queries added.  Initializing with first query.
12:30:15.683 [ESH-safeCall-4] DEBUG o.h.p.w.WindowsCentralProcessor - Initialized Processor
12:30:15.794 [ESH-safeCall-5] DEBUG oshi.util.platform.windows.WmiUtil - Connected to ROOT\CIMV2 WMI namespace
12:30:15.794 [ESH-safeCall-5] DEBUG oshi.util.platform.windows.WmiUtil - Query: SELECT Name,Manufacturer,Model,SerialNumber,Size FROM Win32_DiskDrive 
12:30:15.800 [ESH-safeCall-5] DEBUG oshi.util.platform.windows.WmiUtil - Connected to ROOT\CIMV2 WMI namespace
12:30:15.800 [ESH-safeCall-5] DEBUG oshi.util.platform.windows.WmiUtil - Query: SELECT Name,CommandLine,ExecutionState,ProcessID,ParentProcessId,ThreadCount,Priority,VirtualSize,WorkingSetSize,KernelModeTime,UserModeTime,CreationDate FROM Win32_Process 
12:30:15.983 [ESH-safeCall-4] DEBUG oshi.util.platform.windows.WmiUtil - Connected to ROOT\CIMV2 WMI namespace
12:30:15.983 [ESH-safeCall-4] DEBUG oshi.util.platform.windows.WmiUtil - Query: SELECT Name,Manufacturer,Model,SerialNumber,Size FROM Win32_DiskDrive 
12:30:15.989 [ESH-safeCall-4] DEBUG oshi.util.platform.windows.WmiUtil - Connected to ROOT\CIMV2 WMI namespace
12:30:15.989 [ESH-safeCall-4] DEBUG oshi.util.platform.windows.WmiUtil - Query: SELECT Name,CommandLine,ExecutionState,ProcessID,ParentProcessId,ThreadCount,Priority,VirtualSize,WorkingSetSize,KernelModeTime,UserModeTime,CreationDate FROM Win32_Process 
java.lang.NumberFormatException: For input string: ""
	at java.lang.NumberFormatException.forInputString(Unknown Source) ~[na:1.8.0_65]
	at java.lang.Long.parseLong(Unknown Source) ~[na:1.8.0_65]
	at java.lang.Long.parseLong(Unknown Source) ~[na:1.8.0_65]
	at java.text.DigitList.getLong(Unknown Source) ~[na:1.8.0_65]
	at java.text.DecimalFormat.parse(Unknown Source) ~[na:1.8.0_65]
	at java.text.SimpleDateFormat.subParse(Unknown Source) ~[na:1.8.0_65]
	at java.text.SimpleDateFormat.parse(Unknown Source) ~[na:1.8.0_65]
	at oshi.util.ParseUtil.cimDateTimeToDate(ParseUtil.java:293) ~[oshi-core-2.5-20160522.160709-12.jar:2.5-SNAPSHOT]
	at oshi.util.platform.windows.WmiUtil.enumerateProperties(WmiUtil.java:462) ~[oshi-core-2.5-20160522.160709-12.jar:2.5-SNAPSHOT]
	at oshi.util.platform.windows.WmiUtil.queryWMI(WmiUtil.java:325) ~[oshi-core-2.5-20160522.160709-12.jar:2.5-SNAPSHOT]
	at oshi.util.platform.windows.WmiUtil.selectObjectsFrom(WmiUtil.java:243) ~[oshi-core-2.5-20160522.160709-12.jar:2.5-SNAPSHOT]
       at oshi.hardware.platform.windows.WindowsCentralProcessor.getProcesses(WindowsCentralProcessor.java:395) ~[oshi-core-2.5-20160522.160709-12.jar:2.5-SNAPSHOT]
```

I guess that the CentralProcessor object may not be fully initialized, because a later call to 

```
cpu.getLogicalProcessorCount();
cpu.getPhysicalProcessorCount();
operatingSystem.getFamily();
operatingSystem.getManufacturer();
operatingSystem.getVersion().toString();
```
has resulted in NPE (I do not have any details about this one)

As you can see this is the snapshot from 2016 05 22.

Can I help you with further information ?

**Comment by svilenvul** [comment]
And one more 
```
13:16:02.199 [ESH-safeCall-4] ERROR o.o.b.s.handler.SysteminfoHandler - Can not instantate Systeminfo object!
java.lang.NumberFormatException: multiple points
	at sun.misc.FloatingDecimal.readJavaFormatString(Unknown Source) ~[na:1.8.0_65]
	at sun.misc.FloatingDecimal.parseDouble(Unknown Source) ~[na:1.8.0_65]
	at java.lang.Double.parseDouble(Unknown Source) ~[na:1.8.0_65]
	at java.text.DigitList.getDouble(Unknown Source) ~[na:1.8.0_65]
	at java.text.DecimalFormat.parse(Unknown Source) ~[na:1.8.0_65]
	at java.text.SimpleDateFormat.subParse(Unknown Source) ~[na:1.8.0_65]
	at java.text.SimpleDateFormat.parse(Unknown Source) ~[na:1.8.0_65]
	at oshi.util.ParseUtil.cimDateTimeToDate(ParseUtil.java:293) ~[oshi-core-2.5-20160522.160709-12.jar:2.5-SNAPSHOT]
	at oshi.util.platform.windows.WmiUtil.enumerateProperties(WmiUtil.java:462) ~[oshi-core-2.5-20160522.160709-12.jar:2.5-SNAPSHOT]
	at oshi.util.platform.windows.WmiUtil.queryWMI(WmiUtil.java:325) ~[oshi-core-2.5-20160522.160709-12.jar:2.5-SNAPSHOT]
	at oshi.util.platform.windows.WmiUtil.selectObjectsFrom(WmiUtil.java:243) ~[oshi-core-2.5-20160522.160709-12.jar:2.5-SNAPSHOT]
	at oshi.hardware.platform.windows.WindowsCentralProcessor.getProcesses(WindowsCentralProcessor.java:395) ~[oshi-core-2.5-20160522.160709-12.jar:2.5-SNAPSHOT]
```

**Comment by dbwiddis** [comment]
Well, the exception is because I failed to catch the `NumberFormatException` on an empty string being sent to the cim date parsing.  But the root of the issue is that it's apparently not reading your process information from WMI.

Does a command line call `wmic process get Name,CommandLine,ExecutionState,ProcessID,ParentProcessId,ThreadCount,Priority,VirtualSize,WorkingSetSize,KernelModeTime,UserModeTime,CreationDate` return useful information, particularly in the `CreationDate` field?

**Comment by svilenvul** [comment]
The issue is not 100 percent reproducible. 
Running the command `wmic process get Name,CommandLine,ExecutionState,ProcessID,ParentProcessId,ThreadCount,Priority,VirtualSize,WorkingSetSize,KernelModeTime,UserModeTime,CreationDate` from cmd seems to be ok - e.g ` 20160527152413.537286+180` . 
Could that be some multithreading issue ? I see that the command is called from two threads simultaneously.

**Comment by dbwiddis** [comment]
Could be a process that made the list but has null information because it ended between being indexed and iterated (internally in WMI).  I can probably just catch and ignore the exception if it's a rare thing.

**Comment by dbwiddis** [comment]
No, actually I think you're right. `SimpleDateFormat` is not thread safe.

**Comment by dbwiddis** [comment]
OK, I switched the CreationDate parsing to use the newer Java 8 thread safe date format (using the `threeten` backport to Java 7).  Give the latest snapshot a try and see if it has fixed this issue.

Still on the fence on on whether to add the threeten dependency, or just update the project to Java 8...

**Comment by dbwiddis** [comment]
Have you been able to test this fix?  I'm getting ready to release 2.5 and want to make sure this has been addressed.

**Comment by svilenvul** [comment]
Sorry, @dbwiddis , that I was not able to test it in the last days, but I will try it right now.

**Comment by dbwiddis** [comment]
Great. I already released 2.5 but there's always room for 2.5.1! :-)

**Comment by dbwiddis** [comment]
Since there was no follow up in over 2 years, I assume it was fixed in 2.5.0.   As of version 3.6.0, we no longer use WMI date parsing for Processes.  What version are you using?

**Comment by pheterkubicom** [comment]
Ok, thanks. I assume that the change works fine then, so switching from SimpleDateFormat to the new java.time package in Java 8 is the way to go. =)
